### PR TITLE
[CUSPARSE] Fix triangular solves with matrices used as right-hand sides

### PR DESCRIPTION
**Pull request by amontoison**

### CODE REVIEW

File: lib/cusparse/interfaces.jl [hunk 1]
```diff
@@ -17,7 +17,7 @@ LinearAlgebra.mul!(C::CuVector{T},adjA::Adjoint{<:Any, <:HermOrSym{T,<:CuSparseM
 
 function mm_wrapper(transa::SparseChar, transb::SparseChar, alpha::Number,
                     A::CuSparseMatrix{T}, B::CuMatrix{T}, beta::Number, C::CuMatrix{T}) where {T}
-    if version() < v"10.3.1" && A isa CuSparseMatrixCSR
+    if version() < v"10.3.1" && A isa CuSparseMatrixCSC
         # generic mm! doesn't work on CUDA 10.1 with CSC matrices
```
Comment (amontoison) [inline lib/cusparse/interfaces.jl]: The comment was wrong or not?

File: lib/cusparse/interfaces.jl [hunk 2]
```diff
@@ -34,11 +34,17 @@ LinearAlgebra.mul!(C::CuMatrix{T},A::HermOrSym{<:Number, <:CuSparseMatrix},B::De
 LinearAlgebra.mul!(C::CuMatrix{T},transA::Transpose{<:Any, <:HermOrSym{<:Number, <:CuSparseMatrix}},B::DenseCuMatrix,alpha::Number,beta::Number) where {T} = mm_wrapper('T',alpha,parent(transA),B,beta,C)
 LinearAlgebra.mul!(C::CuMatrix{T},adjA::Adjoint{<:Any, <:HermOrSym{<:Number, <:CuSparseMatrix}},B::DenseCuMatrix,alpha::Number,beta::Number) where {T} = mm_wrapper('C',alpha,parent(adjA),B,beta,C)
 
-Base.:(\)(A::Union{UpperTriangular{T, S},LowerTriangular{T, S}}, B::DenseCuMatrix{T}) where {T<:BlasFloat, S<:AbstractCuSparseMatrix{T}} = sm('N',A,B,'O')
-Base.:(\)(transA::Transpose{T, UpperTriangular{T, S}}, B::DenseCuMatrix{T}) where {T<:BlasFloat, S<:AbstractCuSparseMatrix{T}} = sm('T',parent(transA),B,'O')
-Base.:(\)(transA::Transpose{T, LowerTriangular{T, S}}, B::DenseCuMatrix{T}) where {T<:BlasFloat, S<:AbstractCuSparseMatrix{T}} = sm('T',parent(transA),B,'O')
-Base.:(\)(adjA::Adjoint{T, UpperTriangular{T, S}},B::DenseCuMatrix{T}) where {T<:BlasFloat, S<:AbstractCuSparseMatrix{T}} = sm('C',parent(adjA),B,'O')
-Base.:(\)(adjA::Adjoint{T, LowerTriangular{T, S}},B::DenseCuMatrix{T}) where {T<:BlasFloat, S<:AbstractCuSparseMatrix{T}} = sm('C',parent(adjA),B,'O')
```
Comment (amontoison) [inline lib/cusparse/interfaces.jl:41]: `sm` was removed from `CUDA.jl`.

File: lib/cusparse/interfaces.jl (hunk 1 above)
Comment (maleadt) [reply lib/cusparse/interfaces.jl]: I don't remember. Let CI be the judge!

Comment (maleadt) [reply lib/cusparse/interfaces.jl]: ... and CI failed for 10.1, so I guess the comment was wrong.

Comment (amontoison) [reply lib/cusparse/interfaces.jl]: Ok, I fixed the typo.

**Comment by amontoison** [comment]
I found that yesterday : https://github.com/JuliaGPU/CUDA.jl/blob/master/lib/cusparse/types.jl#L91-L99.
Maybe instead of having a boolean kwarg `unit_diag`, we can change it for a SparseChar `diag` ('U' or 'N') ? It will remove these lines : https://github.com/JuliaGPU/CUDA.jl/pull/550/files#diff-efc807b0787f197e08b9fcf7f5e046a8bdf66409b07cb71e9b42f81ebbb286f9R199-R200 in `sv2!` and `mv2!`

**Comment by maleadt** [comment]
>  Maybe instead of having a boolean kwarg `unit_diag`, we can change it for a SparseChar `diag` ('U' or 'N') ? It will remove these lines : https://github.com/JuliaGPU/CUDA.jl/pull/550/files#diff-efc807b0787f197e08b9fcf7f5e046a8bdf66409b07cb71e9b42f81ebbb286f9R199-R200 in `sv2!` and `mv2!`

Yeah, that seems better indeed. No need to call convert explicitly, `ccall` does that.

**Comment by amontoison** [comment]
Ok, after this pull request I will update `sv2!` and `mv2!` with another PR to take into account this modification.

**Comment by amontoison** [comment]
I prepared a PR that use `U` or `N` values for `diag` kwarg.
I will open it after this one.
https://github.com/amontoison/CUDA.jl/commit/8f74dd7b148a2ed2f37380295246f3e850be4fc3

**Comment by maleadt** [comment]
Thanks. Since I have to touch the Triangular wrappers for 1.6, I'll rebase your changes onto that branch instead of merging it here and dealing with the conflicts later. IIUC the core fix here is support for UnitTriangular?

**Comment by amontoison** [comment]
Yes, it is in the same spirit of #540 .
The only difference is that `Base.\` was not working anymore for any triangular type because `sm` function was removed in the past and the interface was never updated. I fixed that too.

**Comment by maleadt** [comment]
Included in https://github.com/JuliaGPU/CUDA.jl/pull/575.

**Pull request closed without merging**